From a7a955ea7a6c7b4771459960ca438db0c54256bc Mon Sep 17 00:00:00 2001
From: Jeff Moyer <jmoyer@redhat.com>
Date: Fri, 10 Oct 2014 21:03:25 -0400
Subject: [block] blk-mq: use blk_mq_start_hw_queues() when running requeue work

Message-id: <1412975015-5370-25-git-send-email-jmoyer@redhat.com>
Patchwork-id: 97475
O-Subject: [RHEL7 PATCH 24/34] blk-mq: use blk_mq_start_hw_queues() when running requeue work
Bugzilla: 1146660
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Mike Snitzer <snitzer@redhat.com>

This is a backport of the following commit.  This resolves bug
1146660.

  commit 8b95741569eabc5eb17da71d1d3668cdb0bef86c
  Author: Jens Axboe <axboe@fb.com>
  Date:   Fri Sep 19 13:10:29 2014 -0600

    blk-mq: use blk_mq_start_hw_queues() when running requeue work

    When requests are retried due to hw or sw resource shortages,
    we often stop the associated hardware queue. So ensure that we
    restart the queues when running the requeue work, otherwise the
    queue run will be a no-op.

    Signed-off-by: Jens Axboe <axboe@fb.com>

Signed-off-by: Jeff Moyer <jmoyer@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/block/blk-mq.c b/block/blk-mq.c
index a0186c7..2afa800 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -482,7 +482,11 @@ static void blk_mq_requeue_work(struct work_struct *work)
   blk_mq_insert_request(rq, false, false, false);
  }
 
- blk_mq_run_queues(q, false);
+ /*
+  * Use the start variant of queue running here, so that running
+  * the requeue work will kick stopped queues.
+  */
+ blk_mq_start_hw_queues(q);
 }
 
 void blk_mq_add_to_requeue_list(struct request *rq, bool at_head)
-- 
1.7.1